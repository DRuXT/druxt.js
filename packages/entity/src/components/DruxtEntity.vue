<script>
import merge from 'deepmerge'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import DruxtModule from 'druxt/dist/components/DruxtModule.vue'
import { mapActions } from 'vuex'

/**
 * The `<DruxtEntity />` component fetches the Entity JSON:API resource and
 * schema data to render the Entities fields based on the Drupal Display
 * Mode configuration.
 * 
 * Features:
 * - Display mode based field system
 * - Scoped slots
 * - Query settings
 *
 * @example @lang vue
 * <DruxtEntity
 *   type="node--article"
 *   :uuid="uuid"
 *   mode="teaser"
 * />
 */
export default {
  name: 'DruxtEntity',

  extends: DruxtModule,

  /** */
  props: {
    /**
     * Drupal display mode.
     *
     * @type {string}
     * @default default
     */
    mode: {
      type: String,
      default: 'default'
    },

    /**
     * JSON:API resource type.
     *
     * @type {string}
     */
    type: {
      type: String,
      required: true
    },

    /**
     * Entity UUID.
     *
     * @type {string}
     */
    uuid: {
      type: [Boolean, String],
      default: false
    },

    /**
     * Drupal display schema type, 'view' or 'form'.
     * 
     * @type {('view'|'form')}
     */
    schemaType: {
      type: String,
      default: undefined,
    },
  },

  /**
   * Nuxt.js fetch method.
   */
  async fetch() {
    // Fetch Schema.
    try {
      this.schema = await this.getSchema({
        resourceType: this.type,
        mode: this.mode,
        schemaType: this.schemaType || 'view',
      })
    } catch(e) {}

    // Build wrapper component object.
    const options = this.getModuleComponents()
    let component = {
      is: (((options.filter(o => o.global) || [])[0] || {}).name || 'DruxtWrapper'),
      options: options.map(o => o.name) || [],
    }

    // Get wrapper component data to merge with module settings.
    const wrapperData = await this.getWrapperData(component.is)
    component.settings = merge((this.$druxtEntity || {}).options || {}, wrapperData.druxt || {}, { arrayMerge: (dest, src) => src })

    // Fetch Entity resource.
    if (this.uuid && !this.value) {
      const query = this.getQuery(component.settings)
      const entity = (await this.getResource({ type: this.type, id: this.uuid, query })).data
      this.model = JSON.parse(JSON.stringify(entity || {}))
    }

    // Get scoped slots.
    component.slots = Object.keys(this.getScopedSlots())

    // Build wrapper component propsData.
    component = { ...component, ...this.getModulePropsData(wrapperData.props) }

    // Set component data.
    this.component = component
  },

  fetchKey(getCounter) {
    const parts = ['DruxtEntity', this.type, this.uuid, this.mode, this.schemaType].filter((o) => o)
    return [...parts, getCounter(parts.join(':'))].join(':')
  },

  /**
   * @property {object} model - The model object.
   * @property {object} schema - The DruxtSchema object.
   */
  data: ({ type, value }) => ({
    model: {
      attributes: {},
      relationships: {},
      type,
      ...value,
    },
    schema: null,
  }),

  /** */
  computed: {
    /**
     * The Entity object.
     *
     * @return {object}.
     */
    entity: ({ model }) => ({ ...model }),

    /**
     * Entity fields based on Display mode.
     *
     * @return {object}
     */
    fields: ({ errors, isEmpty, model, schema, schemaType }) => {
      if (!schema) return false

      const fields = {}
      for (const field of schema.fields) {
        const relationship = !!((field.settings || {}).storage || {}).target_type || !!(model.relationships || {})[field.id]
        const value = relationship ? ((model || {}).relationships || {})[field.id] : ((model || {}).attributes || {})[field.id]

        // Filter out empty fields if not using the Form schema type.
        // @todo - Make this configurable?
        if (schemaType !== 'form' && isEmpty(value)) continue

        fields[field.id] = {
          id: field.id,
          // @todo - Remove deprecated 'data'.
          data: value,
          errors: (errors || []).filter((o) => ((o.source || {}).pointer || '').startsWith(`/data/attributes/${field.id}`)),
          relationship,
          schema: {
            config: schema.config,
            ...field,
          },
          value,
        }
      }

      return fields
    },
  },

  watch: {
    mode() {
      this.$fetch()
    },

    schemaType() {
      this.$fetch()
    },

    type() {
      this.$fetch()
    },

    uuid() {
      this.$fetch()
    },
  },

  methods: {
    /**
     * Builds the query for the JSON:API request.
     *
     * @param {ModuleSettings} settings - The merged module and component settings object.
     *
     * @return {boolean|object}
     */
    getQuery(settings) {
      const query = new DrupalJsonApiParams()

      // Build fields list.
      const fields = (settings.query || {}).schema
        ? [...((this.schema || {}).fields || []).map((o) => o.id), ...((settings.query || {}).fields || [])]
        : ((settings.query || {}).fields || [])
      if (fields.length) {
        query.addFields(this.type, fields)
        return query
      }
      return false
    },

    /**
     * Checks if an Entity field is empty.
     *
     * @param {*} value - Field value.
     * @return {boolean}
     */
    isEmpty(value) {
      if (typeof value === 'undefined') return true

      if (!value) return true

      if (Array.isArray(value.data) && !value.data.length) return true

      if (typeof value.data !== 'undefined' && !value.data) return true

      return false
    },

    /**
     * Maps Vuex action to methods.
     */
    ...mapActions({
      getResource: 'druxt/getResource',
      getSchema: 'druxtSchema/get'
    })
  },

  /** DruxtModule settings */
  druxt: {
    /**
     * Provides the available component naming options for the Druxt Wrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {ComponentOptions}
     */
    componentOptions: ({ mode, schema, schemaType, type }) => ([
      [(schema || {}).resourceType || type, ((schema || {}).config || {}).mode || mode, ((schema || {}).config || {}).schemaType || schemaType || 'view'],
      [((schema || {}).config || {}).mode || mode],
    ]),

    /**
     * Provides propsData for the DruxtWrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {PropsData}
     */
    propsData: ({ fields, model, schema }) => ({ entity: model, fields, schema, value: model }),

    /**
     * Provides the scoped slots object for the Module render function.
     *
     * A scoped slot is provided for each field being rendered, as per the
     * current display mode.
     * 
     * Additionally, the `default` slot will render all fields as per the
     *
     * @example <caption>DruxtEntity**ResourceType**.vue</caption> @lang vue
     * <template>
     *   <div>
     *     <slot name="content" />
     *     <slot :name="field_name" />
     *   </div>
     * </template>
     *
     * @return {ScopedSlots} The Scoped slots object.
     */
    slots(h) {
      const scopedSlots = {}

      if (!this.schema) {
        if (this.$nuxt.context.isDev) {
          scopedSlots.default = (attrs) => h('details',
            {
              attrs,
              style: {
                border: '2px dashed lightgrey',
                margin: '0.5em 0',
                padding: '1em',
              },
            },
            [
              h('summary', [`[DruxtEntity] Missing schema for '${this.type}--${this.mode}'`]),
              h('br'),
              h('label', ['Component options:', h('ul', this.component.options.map((s) => h('li', [s])))]),
              h('br'),
              h('label', ['Entity:', h('pre', [JSON.stringify(this.entity, null, '  ')])]),
            ]
          )
        }
        return scopedSlots
      }

      // Build scoped slots for each field.
      Object.entries(this.fields).map(([id, field]) => {
        scopedSlots[id] = (attrs) => h('DruxtField', {
          attrs,
          key: id,
          props: field,
          on: {
            input: (value) => {
              const type = !field.relationship ? 'attributes' : 'relationships'
              this.model[type][id] = value
              this.$emit('input', this.model)
            }
          },
          ref: id,
        })
      })

      // Build default slot.
      scopedSlots.default = (attrs) => Object.entries(this.fields).map(([id]) => scopedSlots[id](attrs))

      return scopedSlots
    },
  },
}

/**
 * Provides the available naming options for the Wrapper component.
 *
 * @typedef {array[]} ComponentOptions
 *
 * @example @lang js
 * [
 *   'DruxtEntity[ResourceType][DisplayMode][SchemaType]',
 *   'DruxtEntity[ResourceType][DisplayMode]',
 *   'DruxtEntity[ResourceType]',
 *   'DruxtEntity[DisplayMode]',
 * ]
 *
 * @example <caption>Article Node (default)</caption> @lang js
 * [
 *   'DruxtEntityNodeArticleDefaultView',
 *   'DruxtEntityNodeArticleDefault',
 *   'DruxtEntityNodeArticle',
 *   'DruxtEntityNodeDefault',
 * ]
 */

/**
 * Provides settings for the Entity module, via the `nuxt.config.js` `druxt.entity`
 * or the Wrapper component `druxt` object.
 *
 * @typedef {object} ModuleSettings
 * @param {string[]} fields - An array of fields to filter from the JSON:API Resource.
 * @param {boolean} schema - Whether to automatically detect fields to filter, per the Display mode.
 *
 * @example @lang js
 * {
 *   fields: [],
 *   schema: false,
 * }
 *
 * @example @lang vue
 * <script>
 * export default {
 *   druxt: {
 *     query: {
 *       fields: ['title']
 *       schema: true,
 *     },
 *   }
 * }
 */

/**
 * Provides propsData for use in the Wrapper component.
 *
 * @typedef {object} PropsData
 * @param {object} entity - The Drupal Entity JSON:API resource data.
 * @param {object} fields - Drupal Entity Fields data.
 * @param {object} schema - DruxtJS Schema object.
 * @param {object} value - The Entity value.
 *
 * @example @lang js
 * {
 *   entity: {
 *     attributes: {},
 *     id: '43118086-cca5-4c62-b11e-f1d870050ebd',
 *     links: {},
 *     relationships: {},
 *     type: 'node--article',
 *   },
 *   fields: {
 *     body: {},
 *     field_media_image: {},
 *     field_tags: {},
 *   },
 *   schema: {
 *     config: {},
 *     fields: {},
 *     groups: {},
 *     id: 'node--article--default--view',
 *     resourceType: 'node--article',
 *   },
 *   value: {},
 * }
 */

/**
 * Provides scoped slots for use in the Wrapper component.
 *
 * @typedef {object} ScopedSlots
 * @param {function} * - Slot per field.
 * @param {function} default - All fields per Display mode.
 *
 * @example <caption>DruxtEntity**ResourceType**.vue</caption> @lang vue
 * <template>
 *   <div>
 *     <slot name="content" />
 *     <slot :name="field_name" />
 *   </div>
 * </template>
 */
</script>
