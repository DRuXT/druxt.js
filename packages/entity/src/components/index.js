/**
 * The `<DruxtEntity />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntity
 * @see {@link ./DruxtEntity|DruxtEntity}
 */
export { default as DruxtEntity } from './DruxtEntity.vue'

/**
 * The `<DruxtEntityForm />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntityForm
 * @see {@link ./DruxtEntityForm|DruxtEntityForm}
 */
export { default as DruxtEntityForm } from './DruxtEntityForm.vue'

/**
 * The `<DruxtEntityFormButtons />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtEntityFormButtons
 * @see {@link ./DruxtEntityFormButtons|DruxtEntityFormButtons}
 */
export { default as DruxtEntityFormButtons } from './DruxtEntityFormButtons.vue'

/**
 * The `<DruxtField />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtField
 * @see {@link ./DruxtField|DruxtField}
 */
export { default as DruxtField } from './DruxtField.vue'

/**
 * The `<DruxtRouterEntity />` Vue.js component.
 * 
 * @type {object}
 * @exports DruxtRouterEntity
 * @see {@link ./DruxtRouterEntity|DruxtRouterEntity}
 */
export { default as DruxtRouterEntity } from './DruxtRouterEntity.vue'

export * from './fields'
