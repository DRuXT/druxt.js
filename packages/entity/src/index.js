/**
 * Vue.js mixins.
 *
 * @type {object}
 * @exports DruxtEntityMixins
 * @see {@link ./mixins|DruxtEntityMixins}
 * @see {@link ./mixins/entity|DruxtEntityMixin}
 * @see {@link ./mixins/field|DruxtFieldMixin}
 */
export * from './mixins'

/**
 * Default function to alert user to incorrectly installed module.
 *
 * This was added as part of the @nuxt/kit update due to breaking changes.
 */
export default () => {
  throw new Error("DruxtEntity Nuxt module must be installed as 'druxt-entity/nuxt'")
}
