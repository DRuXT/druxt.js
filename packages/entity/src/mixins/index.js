export * from './context.js'
export * from './componentSuggestion.js'

/**
 * Vue.js mixin for DruxtEntity and DruxtEntityForm Wrapper components.
 *
 * @type {object}
 * @exports DruxtEntityMixin
 * @see {@link ./entity|DruxtEntityMixin}
 */
export { DruxtEntityMixin } from './entity'

/**
 * Vue.js mixin for DruxtField Wrapper components.
 *
 * @type {object}
 * @exports DruxtFieldMixin
 * @see {@link ./field|DruxtFieldMixin}
 */
export { DruxtFieldMixin } from './field'
