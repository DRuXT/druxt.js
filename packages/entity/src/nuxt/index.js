import { defineNuxtModule, installModule } from '@nuxt/kit'
import { join } from 'path'
// import DruxtEntityStorybook from '../nuxtStorybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds dependant modules.
 * - Adds Nuxt plugin.
 * - Adds Nuxt Storybook integration.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-entity'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @todo Document options.
 *
 * @property {object} options.druxt - DruxtJS root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 */
const DruxtEntityNuxtModule = defineNuxtModule({
  meta: {
    name: 'druxt-entity',
  },
  defaults: {
    baseUrl: '',
    endpoint: '/jsonapi'
  },

  async setup(moduleOptions, nuxt) {
    // Set default options.
    const options = {
      baseUrl: moduleOptions.baseUrl,
      ...nuxt.options?.druxt,
      entity: {
        query: {},
        ...nuxt.options?.druxt?.entity,
        ...moduleOptions,
        components: {
          fields: false,
          ...nuxt.options?.druxt?.entity?.components,
          ...moduleOptions?.components
        }
      }
    }

    // Add dependant modules.
    await installModule('druxt/nuxt', options, nuxt)
    await installModule('druxt-schema/nuxt', { baseUrl: options.baseUrl }, nuxt)

    // Register components directories.
    nuxt.hook('components:dirs', dirs => {
      dirs.push({
        path: join(__dirname, '../dist/components'),
        ignore: ['fields']
      })
      if (options.entity.components.fields) {
        dirs.push({ path: join(__dirname, '../dist/components/fields') })
      }
    })

    // Nuxt Storybook.
    // @TODO - @nuxt/kit and @nuxt/storybook aren't compatible.
    // this.nuxt.hook('storybook:config', async ({ stories }) => {
    //   await DruxtEntityStorybook.call(this, { stories })
    // })
  }
})

export default DruxtEntityNuxtModule
