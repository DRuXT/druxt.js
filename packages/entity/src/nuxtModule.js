import { resolve } from 'path'
import DruxtEntityStorybook from './nuxtStorybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds dependant modules.
 * - Adds Nuxt plugin.
 * - Adds Nuxt Storybook integration.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-entity'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @todo Add module level options.
 * @todo Document options.
 *
 * @property {object} options.druxt - DruxtJS root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 */
const DruxtEntityNuxtModule = function (moduleOptions = {}) {
  const { options } = this

  // Use root level Druxt options.
  if (typeof options === 'undefined' || !options.druxt) {
    throw new TypeError('Druxt settings missing.')
  }

  // Add dependant modules.
  const modules = ['druxt', 'druxt-schema'].filter((module) => !options.modules.includes(module))
  for (const module of modules) {
    this.addModule(module)
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/plugin.js'),
    fileName: 'druxt-entity.js',
    options: options.druxt
  })

  // Nuxt Storybook.
  this.nuxt.hook('storybook:config', async ({ stories }) => {
    await DruxtEntityStorybook.call(this, { stories })
  })
}

DruxtEntityNuxtModule.meta = require('../package.json')

export { DruxtEntityNuxtModule }
