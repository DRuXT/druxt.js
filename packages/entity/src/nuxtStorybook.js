import { resolve } from 'path'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { DruxtClient } from 'druxt'

// @todo - Use component naming convention for title.
const titleFn = (parts) =>
  parts.map((part) =>
    part.charAt(0).toUpperCase() + part.slice(1).replace(/_/g, ' ')
  ).join('/')

export default async function ({ stories }) {
  const { addTemplate, options } = this

  const druxt = new DruxtClient(options.druxt.baseUrl, options.druxt)

  // Fetch and process schemas.
  const [view, form] = (await Promise.all(['view', 'form'].map(async (type) =>
    await druxt.getCollectionAll(`entity_${type}_display--entity_${type}_display`, new DrupalJsonApiParams()
      .addFilter('status', 1)
      .addFields(`entity_${type}_display--entity_${type}_display`, [
        'bundle',
        'mode',
        'targetEntityType',
      ])
    )))).map((schema) => schema.map((collection) => collection.data.map((o) => o)).flat())
  const schemas = [...view, ...form]

  // Get Entity type and bundle data.
  const entityTypes = Array.from(new Set(schemas.map((o) => o.attributes.targetEntityType)))
    .map((entity) => ({
      entity,
      bundles: Array.from(new Set(schemas.filter((o) => o.attributes.targetEntityType === entity).map((o) => o.attributes.bundle))),
    })
  )

  // Fetch Entities for UUID controls.
  const entities = await Promise.all([
    ...entityTypes.map(({ entity, bundles }) => bundles.map(async (bundle) => {
      try {
        return {
          resourceType: `${entity}--${bundle}`,
          entities: (await druxt.getCollection(`${entity}--${bundle}`, new DrupalJsonApiParams()
            .addFilter('status', 1)
            .addFields(`${entity}--${bundle}`, ['id', 'title', 'name', 'info'])
          )).data.map((o) => ({
            id: o.id,
            title: (o.attributes || {}).title || (o.attributes || {}).name || (o.attributes || {}).info || o.id,
          })),
        }
      }
      catch(e) {
        return {
          resourceType: `${entity}--${bundle}`,
          entities: [],
        }
      }
    })).flat()
  ])

  // Build Entity Stories.
  const entityStories = entityTypes.map(({ entity, bundles }) => bundles.map((bundle) => {
    return Object.entries({ view, form }).map(([type, schemas]) => {
      const displays = Array.from(new Set(schemas.filter((o) => o.attributes.targetEntityType === entity && o.attributes.bundle === bundle).map((o) => o.attributes.mode).sort((a) => (a === 'default' ? -1 : 0))))
      if (!displays) {
        return false
      }

      const resourceType = `${entity}--${bundle}`
      const component = type === 'view' ? 'druxt-entity' : 'druxt-entity-form'

      addTemplate({
        src: resolve(__dirname, `../nuxt/${component}.stories.js`),
        fileName: `stories/${component}.${resourceType}.stories.js`,
        options: {
          displays,
          entities: entities.find((o) => o.resourceType === resourceType).entities,
          resourceType,
          title: titleFn(['Druxt Entity', entity, bundle, `${type} displays`]),
        },
      })
  
      return resolve(options.buildDir, `./stories/${component}.${resourceType}.stories.js`)
    })
  })).flat(2)

  stories.push.apply(stories, entityStories)
}
