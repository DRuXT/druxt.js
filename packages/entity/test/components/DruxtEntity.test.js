import 'regenerator-runtime/runtime'
import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import mockAxios from 'jest-mock-axios'
import { getMockResource } from '../../../test-utils/src'

import { DruxtClient, DruxtStore } from '../../../druxt/src'
import { DruxtSchemaStore } from '../../../schema/src'
import DruxtEntity from '../../src/components/DruxtEntity.vue'
import DruxtField from '../../src/components/DruxtField.vue'

let localVue

const stubs = ['DruxtDebug', 'DruxtEntityNodePage']
let store

const mocks = {
  $druxt: {
    settings: {},
  },
  $fetchState: {
    pending: false
  },
  $nuxt: {
    context: {
      isDev: false,
    },
  },
}

const mountComponent = (propsData) => {
  return mount(DruxtEntity, { localVue, mocks, propsData, store, stubs })
}

describe('DruxtEntity', () => {
  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(Vuex)
    localVue.component('DruxtField', DruxtField)

    mockAxios.reset()

    // Setup vuex store.
    store = new Vuex.Store()

    DruxtStore({ store })
    store.$druxt = new DruxtClient('https://demo-api.druxtjs.org')

    DruxtSchemaStore({ store })
    store.$druxtSchema = {
      import: (schema) => {
        return require(`../../../../test/__fixtures__/schemas/${schema}.json`)
      }
    }

    store.app = { context: { error: jest.fn() }, store }
  })

  test('node--page', async () => {
    const mockPage = await getMockResource('node--page')

    const wrapper = mountComponent({ uuid: mockPage.data.id, type: 'node--page' })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtEntity.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe(`DruxtEntity:node--page:${mockPage.data.id}:default:0`)

    expect(mockAxios.get).toHaveBeenCalledTimes(2)

    // Props.
    expect(wrapper.vm.mode).toBe('default')
    expect(wrapper.vm.type).toBe('node--page')
    expect(wrapper.vm.uuid).toBe(mockPage.data.id)

    // Data.
    expect(Object.keys(wrapper.vm.component.$attrs)).toStrictEqual([
      'entity', 'fields', 'schema', 'value'
    ])
    expect(wrapper.vm.component.is).toBe('DruxtEntityNodePage')
    expect(wrapper.vm.component.options).toStrictEqual([
      'DruxtEntityNodePageDefaultView',
      'DruxtEntityNodePageDefault',
      'DruxtEntityNodePage',
      'DruxtEntityDefault',
    ])
    expect(wrapper.vm.component.props).toStrictEqual({})
    expect(Object.keys(wrapper.vm.component.propsData)).toStrictEqual([
      'entity', 'fields', 'schema', 'value',
    ])

    expect(Object.keys(wrapper.vm.entity)).toStrictEqual(['type', 'id', 'links', 'attributes', 'relationships'])
    expect(Object.keys(wrapper.vm.fields)).toStrictEqual(['body'])
    expect(Object.keys(wrapper.vm.schema)).toStrictEqual([
      'id', 'resourceType', 'fields', 'groups', 'config'
    ])

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('node--page - filtered', async () => {
    const mockPage = await getMockResource('node--page')

    localVue.component('DruxtEntityNodePageDefault', {
      props: ['entity', 'fields', 'schema', 'value'],
      druxt: {
        query: {
          fields: ['title'],
          schema: true,
        },
      },
      render(h) {
        return h('div', [JSON.stringify(this.entity)])
      }
    })

    const wrapper = mountComponent({ uuid: mockPage.data.id, type: 'node--page' })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Data.
    expect(wrapper.vm.component.$attrs).toStrictEqual({})
    expect(wrapper.vm.component.is).toBe('DruxtEntityNodePageDefault')
    expect(Object.keys(wrapper.vm.component.props)).toStrictEqual([
      'value', 'schema', 'fields', 'entity'
    ])

    expect(Object.keys(wrapper.vm.entity)).toStrictEqual(['type', 'id', 'links', 'attributes'])

    // Methods.
    expect(wrapper.vm.getQuery(wrapper.vm.component.settings).data.fields['node--page']).toBe('body,links,content_moderation_control,title')
  })

  test('node--page - include', async () => {
    const mockPage = await getMockResource('node--page')

    localVue.component('DruxtEntityNodePageDefault', {
      props: ['entity', 'fields', 'schema', 'value'],
      druxt: {
        query: {
          fields: [['title'], ['user--user', ['display_name']]],
          include: ['uid'],
          schema: true,
        },
      },
      render(h) {
        return h('div', [JSON.stringify(this.entity)])
      }
    })

    const wrapper = mountComponent({ uuid: mockPage.data.id, type: 'node--page' })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Data.
    expect(wrapper.vm.component.$attrs).toStrictEqual({})
    expect(wrapper.vm.component.is).toBe('DruxtEntityNodePageDefault')
    expect(Object.keys(wrapper.vm.component.props)).toStrictEqual([
      'value', 'schema', 'fields', 'entity'
    ])

    expect(Object.keys(wrapper.vm.entity)).toStrictEqual(['type', 'id', 'links', 'attributes', 'relationships', 'included'])
    expect(Object.keys(wrapper.vm.entity.included[0])).toStrictEqual([
      'type', 'id', 'links', 'attributes'
    ])

    // Methods.
    expect(wrapper.vm.getQuery(wrapper.vm.component.settings).data.fields['node--page']).toBe('body,links,content_moderation_control,title')
  })


  test('v-model - entity', async () => {
    const model = { attributes: {}, included: undefined, relationships: {}, type: 'node--page' }
    const Component = {
      template: "<DruxtEntity v-model='model' :type='model.type' ref='component' />",
      components: { DruxtEntity },
      data: () => ({ model }),
    }
    const wrapper = mount(Component, { localVue, mocks, store })
    expect(wrapper.vm.$refs.component.entity).toStrictEqual(model)
    expect(wrapper.vm.$refs.component.type).toStrictEqual(model.type)
    expect(wrapper.vm.$refs.component.value).toStrictEqual(model)

    expect(mockAxios.get).toHaveBeenCalledTimes(0)

    const mockData = {
      ...model,
      attributes: {
        body: {
          value: 'Test'
        }
      }
    }

    wrapper.vm.$refs.component.$emit('input', mockData)
    expect(wrapper.vm.model).toStrictEqual(mockData)
  })

  test('v-model - fields', async () => {
    const mockPage = await getMockResource('node--page')

    const propsData = { uuid: mockPage.data.id, type: 'node--page' }
    const wrapper = mount(DruxtEntity, { localVue, mocks, propsData, store })

    await wrapper.vm.$options.fetch.call(wrapper.vm)

    wrapper.vm.$refs.body.$emit('input', 'test')
    expect(wrapper.vm.model.attributes.body).toBe('test')
  })

  test('watch - props $fetch', async () => {
    const $fetch = jest.fn()
    expect($fetch).toHaveBeenCalledTimes(0)
    DruxtEntity.watch.mode.call({ $fetch })
    expect($fetch).toHaveBeenCalledTimes(1)
    DruxtEntity.watch.schemaType.call({ $fetch })
    expect($fetch).toHaveBeenCalledTimes(2)
    DruxtEntity.watch.type.call({ $fetch })
    expect($fetch).toHaveBeenCalledTimes(3)
    DruxtEntity.watch.uuid.call({ $fetch })
    expect($fetch).toHaveBeenCalledTimes(4)
  })

  test('missing schema', async () => {
    mocks.$nuxt.context.isDev = true
    // TODO : Update test to use { getMockResource } from 'druxt-test-utils'?
    const wrapper = mountComponent({ uuid: '7adbf02c-6e41-40ae-9124-8b4781f9c160', type: 'file--file' })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.schema).toBe(null)
    expect(wrapper.html()).toMatchSnapshot()
  })

  // test('deprecated', () => {
  //   expect(DruxtEntity.methods.isEmpty()).toBe(true)
  //   expect(DruxtEntity.methods.isEmpty(false)).toBe(true)
  //   expect(DruxtEntity.methods.isEmpty({ data: [] })).toBe(true)
  //   expect(DruxtEntity.methods.isEmpty({ data: false })).toBe(true)
  //   expect(DruxtEntity.methods.isEmpty({ data: [{}] })).toBe(false)
  // })
})
