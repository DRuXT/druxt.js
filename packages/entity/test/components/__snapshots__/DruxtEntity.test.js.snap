// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtEntity missing schema 1`] = `
"<div>
  <div entity=\\"[object Object]\\" value=\\"[object Object]\\">
    <details style=\\"border: 2px dashed lightgrey; margin: 0.5em 0px; padding: 1em;\\">
      <summary>[DruxtEntity] Missing schema for 'file--file--default'</summary><br><label>Component options:<ul>
          <li>DruxtEntityFileFileDefaultView</li>
          <li>DruxtEntityFileFileDefault</li>
          <li>DruxtEntityFileFile</li>
          <li>DruxtEntityDefault</li>
        </ul></label><br><label>Entity:<pre>{
  \\"type\\": \\"file--file\\",
  \\"id\\": \\"7adbf02c-6e41-40ae-9124-8b4781f9c160\\",
  \\"links\\": {
    \\"self\\": {
      \\"href\\": \\"https://demo-api.druxtjs.org/en/jsonapi/file/file/7adbf02c-6e41-40ae-9124-8b4781f9c160\\"
    }
  },
  \\"attributes\\": {
    \\"drupal_internal__fid\\": 3,
    \\"langcode\\": \\"en\\",
    \\"filename\\": \\"vegan-chocolate-nut-brownies.jpg\\",
    \\"uri\\": {
      \\"value\\": \\"public://vegan-chocolate-nut-brownies.jpg\\",
      \\"url\\": \\"/sites/default/files/vegan-chocolate-nut-brownies.jpg\\"
    },
    \\"filemime\\": \\"image/jpeg\\",
    \\"filesize\\": 109982,
    \\"status\\": true,
    \\"created\\": \\"2021-05-01T06:24:09+00:00\\",
    \\"changed\\": \\"2021-05-01T06:24:09+00:00\\"
  },
  \\"relationships\\": {
    \\"uid\\": {
      \\"data\\": null,
      \\"links\\": {
        \\"related\\": {
          \\"href\\": \\"https://demo-api.druxtjs.org/en/jsonapi/file/file/7adbf02c-6e41-40ae-9124-8b4781f9c160/uid\\"
        },
        \\"self\\": {
          \\"href\\": \\"https://demo-api.druxtjs.org/en/jsonapi/file/file/7adbf02c-6e41-40ae-9124-8b4781f9c160/relationships/uid\\"
        }
      }
    }
  }
}</pre></label>
    </details>
  </div>
</div>"
`;

exports[`DruxtEntity node--page 1`] = `
"<div>
  <druxtentitynodepage-stub entity=\\"[object Object]\\" fields=\\"[object Object]\\" schema=\\"[object Object]\\" value=\\"[object Object]\\"></druxtentitynodepage-stub>
</div>"
`;
